core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommit.java

```diff
@@ -127,14 +139,63 @@ public void delete(List<Delete> deletes) throws CrudException {
     }
   }
 
+  @Override
+  public void insert(Insert insert) throws CrudException {
+    insert = copyAndSetTargetToIfNot(insert);
+    Put put = ConsensusCommitUtils.createPutForInsert(insert);
+    checkMutation(put);
+    crud.put(put);
+  }
+
+  @Override
+  public void upsert(Upsert upsert) throws CrudException {
+    upsert = copyAndSetTargetToIfNot(upsert);
+    Put put = ConsensusCommitUtils.createPutForUpsert(upsert);
+    checkMutation(put);
+    try {
+      crud.put(put);
+    } catch (UncommittedRecordException e) {
+      lazyRecovery(e);
+      throw e;
+    }
+  }
+
+  @Override
+  public void update(Update update) throws CrudException {
+    update = copyAndSetTargetToIfNot(update);
+    ScalarDbUtils.checkUpdate(update);
+    Put put = ConsensusCommitUtils.createPutForUpdate(update);
+    checkMutation(put);
+    try {
+      crud.put(put);
+    } catch (UnsatisfiedConditionException e) {
+      if (update.getCondition().isPresent()) {
```

Comment:
If the target record specified with the keys doesn't exist even with any condition, I think `RecordNotFoundException` might be more consistent. What do you think?

Reply:
Actually, in Consensus Commit, we can identify whether the condition is not satisfied or the record does not exist if some condition is specified. However, in JDBC transactions, we cannot. This is because for Update with a condition, JDBC transactions execute a `Put` operation with the same condition of the `Update` operation. In this case, `NoMutationExecutionException` is thrown in both cases if the condition is not satisfied or the record does not exist. For more details, please see `JdbcTransaction.java`.

Reply:
Thanks for providing the context. Understood.

It seems there is a bit of inconsistency in that this method doesn't throw `RecordNotFoundException` even if the target record doesn't exist with a condition, while the method throws `RecordNotFoundException` without a condition. I guess it might be confusing and complicate error handling. I'm curious if there is a specific benefit to introduce `RecordNotFoundException` beyond the inconsistency. What do you think?

Reply:
> I'm curious if there is a specific benefit to introduce RecordNotFoundException beyond the inconsistency.

Instead of focusing on a specific benefit, consider what exception should be thrown when users don't specify a condition for Update and the target record doesn't exist. Should we also throw `UnsatisfiedConditionException` in such cases?

I think there are three situations throwing an exception:

1. When a condition is specified for Update:
    1. The target record doesn't exist
    2. The condition isn't satisfied
2. When no condition is specified:
    1. The target record doesn't exist

However, we can't differentiate between whether the condition isn't satisfied (1-i) or the target record doesn't exist (1-ii) when a condition is specified for Update. Therefore, we throw `UnsatisfiedConditionException` in those cases.

The problem is what to throw in situation 2-i. Currently, the code throws `RecordNotFoundException`, but honestly I didn't want to introduce a new exception. Do you have any ideas on this?

Reply:
I think we have some options as follows. Let me order them in my preference:

1. This method always throws `RecordNotFoundException` whenever the target record doesn't exist even with a condition
2. This method always throws `UnsatisfiedConditionException` for simplicity and consistency. Considering only the case that uses no condition, `RecordNotFoundException` would be useful. But with a condition, users need to check the existence of the target record by themselves after catching `UnsatisfiedConditionException`. So, consistent error handling by checking the target record existence by users might be better. `RecordNotFoundException` is removed
3. The current behavior in this PR

For 1, as you explained, it doesn't work considering JDBC Transaction. So, option#2? I don't think `RecordNotFoundException` is so helpful for users, although I'm missing something.

Reply:
I'm concerned that users might be confused if they receive an `UnsatisfiedConditionException` even when they don't specify any condition with the Update operation. That's why I introduced `RecordNotFoundException`. Considering this, I'm not sure if option#2 is a good idea. What do you think?

Reply:
By the way, when users specify a condition with the Put or Delete operation, `UnsatisfiedConditionException` is thrown when the condition is not satisfied or the target record doesn't exist. So this existing behavior is the same when a condition is specified for Update.

Reply:
> I'm concerned that users might be confused if they receive an UnsatisfiedConditionException even when they don't specify any condition with the Update operation

Good point. Indeed, the name of exception sounds a bit confusing. Always throwing `RecordNotFoundException` whose description is like "The target record is not found. This exception is also thrown when not satisfying a given condition" in the option 2 might be better.

> So this existing behavior is the same when a condition is specified for Update.

Yeah, my understanding aligns with this.

I think it's important to prevent users from being surprised with the different exceptions (`RecordNotFoundException` and `UnsatisfiedConditionException`) when the target record doesn't exist, and option 2 would be okay. So throwing `RecordSatisfyingConditionNotFoundException` when a condition is specified would be another option? This is almost same as the current PR's behavior.

1. When a condition is specified for Update:
  i. The target record doesn't exist ===> `RecordSatisfyingConditionNotFoundException`
  ii. The condition isn't satisfied ===> `RecordSatisfyingConditionNotFoundException`
2. When no condition is specified:
  i. The target record doesn't exist ===> `RecordNotFoundException`

I hope users think `RecordSatisfyingConditionNotFoundException` is a kind of extension of `RecordNotFoundException`, at least they're in the same category.

Reply:
Umm, but, throwing `RecordSatisfyingConditionNotFoundException` in Update is not consistent with Put and Delete? As I mentioned, in Put and Delete, `UnsatisfiedConditionException` is thrown when the condition is not satisfied or the target record doesn't exist.

The current situation is as follows:

For Put or Delete:

1. When a condition is specified for Put or Delete:
  i. The target record doesn't exist ===> `UnsatisfiedConditionException`
  ii. The condition isn't satisfied ===> `UnsatisfiedConditionException`
2. When no condition is specified ===> No exception is thrown in any cases

For Update:

1. When a condition is specified for Update:
  i. The target record doesn't exist ===> `UnsatisfiedConditionException`
  ii. The condition isn't satisfied ===> `UnsatisfiedConditionException`
2. When no condition is specified:
  i. The target record doesn't exist ===> `RecordNotFoundException`

What do you think?

Reply:
> throwing RecordSatisfyingConditionNotFoundException in Update is not consistent with Put and Delete?

Ah, sorry. I didn't consider the consistency between Put (and Delete) and Update since Update is newly added. If we need to keep the consistency, yeah, `UnsatisfiedConditionException` still needs to be used?

Also, I thought the existing Put and Delete also throw `UnsatisfiedConditionException` when no condition is specified. But it was the wrong idea. The original option 2 which always throws `UnsatisfiedConditionException` won't be chosen as far as the consistency between Put and Update needs to be kept.

So, the room to decide is whether to throw `RecordNotFoundException` from Update when no condition is specified, I guess. It seems Update can follow the behavior of Put and Delete without throwing the new exception and I didn't get a strong reason to introduce the new one. But I don't have a strong opinion on it.

Reply:
I think one possible design is as follows:

In the current design in this PR, we can only specify `UpdateIf` for the Update operation, but I think we can also introduce `UpdateIfExists`. In this case, the exceptions are thrown as follows:

1. When `UpdateIf` is specified for Update:
  i. If the target record doesn't exist ===> `UnsatisfiedConditionException`
  ii. If the condition isn't satisfied ===> `UnsatisfiedConditionException`
2. When `UpdateIfExists` is specified for Update:
  i. If the target record doesn't exist (and the condition isn't satisfied) ===> `UnsatisfiedConditionException`
3. When no condition is specified ===> No exceptions are thrown in any case

With this approach, we don't need to introduce any new exceptions. What do you think?

BTW, the accurate scenarios for Put and Delete are as follows:

For Put:

1. When `PutIf` is specified for Put:
  i. If the target record doesn't exist ===> `UnsatisfiedConditionException`
  ii. If the condition isn't satisfied ===> `UnsatisfiedConditionException`
2. When `PutIfExists` is specified for Put:
  i. If the target record doesn't exist (and the condition isn't satisfied) ===> `UnsatisfiedConditionException`
3. When `PutIfNotExists` is specified for Put:
  i. If the target record exist (and the condition isn't satisfied) ===> `UnsatisfiedConditionException`
4. When no condition is specified ===> No exceptions are thrown in any case

For Delete:

1. When `DeleteIf` is specified for Delete:
  i. If the target record doesn't exist ===> `UnsatisfiedConditionException`
  ii. If the condition isn't satisfied ===> `UnsatisfiedConditionException`
2. When `DeleteIfExists` is specified for Delete:
  i. If the target record doesn't exist (and the condition isn't satisfied) ===> `UnsatisfiedConditionException`
3. When no condition is specified ===> No exceptions are thrown in any case

Reply:
> I think we can also introduce UpdateIfExists

It seems reasonable and consistent to me! With this explicit condition, throwing `UnsatisfiedConditionException` won't surprise users.

Also, thank you for letting me know the detailed conditions behavior of Put and Delete. I still had a few wrong ideas in the previous comment...
I think introducing `UpdateIf` condition makes Update more consistent with Put and Delete.